src/main/scala/io/iohk/ethereum/merklepatriciatrie/Node.scala

```diff
@@ -0,0 +1,102 @@
+package io.iohk.ethereum.merklepatriciatrie
+
+import io.iohk.ethereum.rlp.{decode => decodeRLP, encode => encodeRLP, _}
+import io.iohk.ethereum.merklepatriciatrie.MerklePatriciaTrie.HashFn
+
+/**
+  * Trie elements
+  */
+private[merklepatriciatrie] sealed trait Node {
+
+  import MerklePatriciaTrie._
+
+  private var hashedCache: Option[Array[Byte]] = None
```

Comment:
Hash does not change but it might be used several times within a single put / remove operation, that's why the its being _cached_.

Hash function is defined within the Trie so you cannot change the hashFn. By cannot I mean that if you do so, your tree will be unusable because you won't be able to access nodes stored in DataSource (as you won't be able to fetch nodes).
MPT uses Keccak256 but it might be useful to use it with another hashFn. Maybe the generalization makes everything more complex and unclear and we should remove hashFn as a trie parameter.

About thread safety, yes, I understand your point (we cannot be sure this optimization will be used as it is) . Maybe a `lazy val` makes more sense but as long as `hashFn` is a parameter we cannot do so.